containertest/containertest.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package containertest provides an ephemeral container (such as a database) for integration testing.
+// It's designed to be used in code that needs to work inside and outside google.
+package containertest
+
+import "io"
+
+// ConnInfo specifies how connect to the created container.
+type ConnInfo struct {
+	Host string
+
+	// PortMapper maps from container port to host port. Do not use after container is closed.
+	PortMapper func(containerPort string) (hostPort string)
+
+	// io.Closer for closing the connection. Should always be initialized, either
+	// with an actual closer or io.NoopCloser.
+	closer io.Closer
+}
+
+// Close implements io.Closer by passing to internal closer field.
+func (c ConnInfo) Close() error {
+	return c.closer.Close()
+}
+
+// Service provides information about what container image should be started and
+// how to know when it has finished stating up.
+type Service interface {
+	// ImageRepository returns the repository for docker image (ex: mysql).
+	ImageRepository() string
+
+	// ImageTag returns the tag for docker image (ex: 5.3).
+	ImageTag() string
+
+	// Environment returns variables to be set in container. Each element is in format of "KEY=VALUE".
+	Environment() []string
+
+	// StartupPorts is the list of ports that must be exposed by container before TestConn is run.
+	StartupPorts() []string
+
+	// TestConn takes a logger and a struct with connection info, and returns nil if app has started.
+	TestConn(progressLogger Logger, info ConnInfo) error
+}
+
+// MustStart starts a container, or panics if there was an error.
+func MustStart(service Service, opts ...Option) ConnInfo {
```

Comment:
It still feels like a really strange API to pass in a structure like this instead of calling `NewMySQL(opts ...Option)`.

Reply:
I was matching the existing API, but woudn't this be more extensible?

Anyone can define their own `Service` outside of `containertest` and still use this API.

Reply:
So if I wanted to make a `Redis` I could do so without having to modify `pkg`